### DISCUSSION TITLE
Collecting multiple directories with the same name in one process

### DISCUSSION THREAD

**Discussion by Lehmann-Fabian** [nextflow-io/nextflow#1998] [General]
Hello there, I have a workflow, which follows this structure:

```groovy
elements = Channel.from("a","b","c").flatten()

def extractDirectory = { it.parent.toString().substring(it.parent.toString().lastIndexOf('/') + 1 ) }

process firstProcess {

    input: 
    val value from elements
    
    output:
    file ("**") into filesInDir
    
    """
    mkdir t u v
    touch t/${value}.txt
    touch u/${value}.txt
    touch v/${value}.txt
    ls
    """

}

filesByDir = filesInDir.flatten().map{ [extractDirectory(it), it] }.buffer( size: Integer.MAX_VALUE, remainder: true )

process secondProcess {

    echo true

    input: 
    tuple val(folderName), file("$folderName/*") from filesByDir
    
    """
    find . -name "*"
    """

}
```

In the secondProcess, I need to stage all the files, and preserve the directory structure. This process should be a single tasks that collects all data from before. In this example, following files should be staged:

t/a.txt
t/b.txt
t/c.txt
u/a.txt
u/b.txt
u/c.txt
v/a.txt
v/b.txt
v/c.txt

With another variant (see below), I get the following error:

> Process `secondProcess` input file name collision -- There are multiple input files for each of the following file names: local-dir/v, local-dir/t, local-dir/u

```groovy
elements = Channel.from("a","b","c").flatten()

def extractDirectory = { it.parent.toString().substring(it.parent.toString().lastIndexOf('/') + 1 ) }

process firstProcess {

    input:
    val value from elements
   
    output:
    file ("{t,u,v}") into filesInDir
   
    """
    mkdir t u v
    touch t/${value}.txt
    touch u/${value}.txt
    touch v/${value}.txt
    ls
    """
}


process secondProcess {
    echo true
    input:
    file("local-dir/*") from filesInDir.flatten().collect()
   
    """
    find local-dir -name "*"
    """
}
```

I’m really looking forward for any suggestions.
Thank you in advance!

**Comment by pditommaso** [comment]
So basically you want to stage all files in the `u` together in the `u` directory, same for the other. The ` secondProcess` should be executed once collecting all previous outputs?

**Reply by Lehmann-Fabian** [reply]
Yes, that is exactly what I need.

**Reply by pditommaso** [reply]
The input having the same name can be renamed or the file name matter?

**Reply by Lehmann-Fabian** [reply]
Unfortunately, the file and directory name cannot be changed. As a workaround, I currently rename the files in the firstProcess and revert that in the secondProcess.

```
#rename the files
results=`find . -name '**.txt*'`
for path in \$results; do
   mv \$path \${path%/*}_\${path##*/}
done;
```

**Reply by pditommaso** [reply]
and how are expecting to name those conflicting files in the name directory?

**Reply by Lehmann-Fabian** [reply]
There are no conflicts, the files within a specific directory have all different names.

**Reply by pditommaso** [reply]
I don't think there's a perfect solution. One workaround could be to change the input in the second process as 

```
process secondProcess {
    echo true
    input:
    file("local-dir/sub-*/*") from filesInDir.flatten().collect()
   
    """
    find local-dir -name "*"
    """
}
```

Then prefixing the command script with a snippet scanning all sub dirs with local-dir re-creating the structure you are expecting.

**Reply by Lehmann-Fabian** [reply]
Hello Paolo, thanks a lot for your proposed solution. The problem with that solution is, if I move files into a new directory I will lose them in the original one, as symbolic links are used. This would prevent me from using the resume function in another run. Actually, I could also create symlinks, instead of moving the files. However, this solution does not bring me many advantages over renaming the files, as I do it now. 

Doesn’t Nextflow support a list of tuples, like this:

```
[[t, ../t/a.txt], 
[u, ../u/a.txt], 
[v, ../v/a.txt], 
[t, ../t/b.txt], 
[u, ../u/b.txt], 
[v, ../v/b.txt], 
[t, ../t/c.txt], 
[u, ../u/c.txt], 
[v, ../v/c.txt]]
```

And map it for each tuple individually, if I use this: 
```tuple val(folderName), file("$folderName/*") from filesByDir```
That would be very helpful, not only in my special case.

**Reply by pditommaso** [reply]
That is possible with the following: 


```
elements = Channel.from("a","b","c").flatten()

def extractDirectory = { it.parent.toString().substring(it.parent.toString().lastIndexOf('/') + 1 ) }

process firstProcess {

    input:
    val value from elements
   
    output:
    file ("{t,u,v}/*.txt") into filesInDir
   
    """
    mkdir t u v
    touch t/${value}.txt
    touch u/${value}.txt
    touch v/${value}.txt
    ls
    """
}


filesGroupped = filesInDir
    .flatten()
    .map { file-> [file.parent.name, file] }
    .groupTuple()


process secondProcess {
    echo true
    input:
    tuple val(id), path("$id/*") from filesGroupped
   
    """
    find $id -name "*"
    """
}

```

The problem is that it will run a task for each group, instead, my understanding is that you need to process all together

**Reply by Lehmann-Fabian** [reply]
You are right, this is exactly what I need, but in a single process. 
I will continue to rename the files in the first process, and rename them again in the second one.
Thank you very much.

**Reply by ysliaw01** [reply]
You can just zip/tar the directory, and pass the zip/tar file as output to your 2nd process. Then unzip/untar the zip/tar file.

**Reply by Lehmann-Fabian** [reply]
Thank you, @ysliaw01, for your feedback. Unfortunately tar/zip is not an option as it would generate a lot of I/O - one directory contains multiple gigabytes of data.

**Reply by Lehmann-Fabian** [reply]
I proposed a solution for the named issue here: https://github.com/nextflow-io/nextflow/pull/2622
